File: 0054-spiral-matrix/0054-spiral-matrix.py
class Solution(object):
    def spiralOrder(self, mat):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        t = l = 0
        r = len(mat[0])
        b = len(mat) 
        res = []
        dr, dc = 0 , 1
        i = j = 0
        while l < r and t < b:
            #left -> right
            for i in range(l,r):
                res.append(mat[t][i])
            t+=1
            #top -> bot
            for i in range(t, b):
                res.append(mat[i][r-1])
            r-=1

            if not (l<r and t<b):
                break

            #right -> left
            for i in range(r-1,l-1,-1):
                res.append(mat[b-1][i])
            b-=1
            #bot -> top
            for i in range(b-1,t-1,-1):
                res.append(mat[i][l])
            l+=1
        
        return res

File: 0054-spiral-matrix/0054-spiral-matrix.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        int R = matrix.length;
        int C = matrix[0].length;
        List<Integer> result = new ArrayList<>();
        
        int l = 0, r = C - 1, u = 0, d = R - 1;
        
        while (l <= r && u <= d) {
            // Traverse from left to right
            for (int i = l; i <= r; i++) {
                result.add(matrix[u][i]);
            }
            u++;
            
            // Traverse from top to bottom
            for (int i = u; i <= d; i++) {
                result.add(matrix[i][r]);
            }
            r--;
            
            // Check if the current boundary is valid
            if (l > r || u > d) {
                break;
            }
            
            // Traverse from right to left
            for (int i = r; i >= l; i--) {
                result.add(matrix[d][i]);
            }
            d--;
            
            // Traverse from bottom to top
            for (int i = d; i >= u; i--) {
                result.add(matrix[i][l]);
            }
            l++;
        }
        
        return result;
    }
}
